modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/richtext/model/BasicTextModel.java

```diff
@@ -0,0 +1,290 @@
+         * @return true if editable
+         */
+        public boolean isUserEditable();
+    }
+
+    private final Content content;
+
+    /**
+     * Constructs an empty model with the specified {@code Content}.
+     * @param c the content to use
+     */
+    public BasicTextModel(Content c) {
+        this.content = c;
+        registerDataFormatHandler(PlainTextFormatHandler.INSTANCE, true, true, 0);
+    }
+
+    /**
+     * Constructs an empty model with the in-memory {@code Content}.
+     */
+    public BasicTextModel() {
+        this(new InMemoryContent());
+    }
+
+    /**
+     * Inserts text at the specified position.
+     * This is a convenience shortcut for {@link #replace(StyleResolver, TextPos, TextPos, String, boolean)}.
+     * @param p the insertion position
+     * @param text the text to insert
+     */
+    public void insertText(TextPos p, String text) {
+        replace(null, p, p, text, false);
+    }
+
+    @Override
+    public int size() {
+        return content.size();
+    }
+
+    @Override
+    public String getPlainText(int index) {
+        return content.getText(index);
+    }
+
+    @Override
+    public RichParagraph getParagraph(int index) {
+        String text = getPlainText(index);
+        return RichParagraph.builder().
+            addSegment(text).
+            build();
+    }
+
+    /**
+     * Determines whether the model is user-editable.
+     * <p>
+     * This method calls {@link BasicTextModel.Content#isUserEditable()}.
+     *
+     * @return true if the model is user-editable
+     */
+    @Override
+    public final boolean isUserEditable() {
+        return content.isUserEditable();
+    }
+
+    @Override
+    protected int insertTextSegment(int index, int offset, String text, StyleAttributeMap attrs) {
+        return content.insertTextSegment(index, offset, text, attrs);
+    }
+
+    @Override
+    protected void insertLineBreak(int index, int offset) {
+        content.insertLineBreak(index, offset);
+    }
+
+    @Override
+    protected void removeRange(TextPos start, TextPos end) {
+        content.removeRange(start, end);
+    }
+
+    @Override
+    protected void insertParagraph(int index, Supplier<Region> generator) {
+        // no-op
+    }
+
+    @Override
+    public StyleAttributeMap getStyleAttributeMap(StyleResolver resolver, TextPos pos) {
+        return StyleAttributeMap.EMPTY;
+    }
+
+    @Override
+    protected final void setParagraphStyle(int index, StyleAttributeMap a) {
+        // no-op
+    }
+
+    @Override
+    protected final void applyStyle(int index, int start, int end, StyleAttributeMap a, boolean merge) {
+        // no-op
+    }
+
+    /**
+     * This content provides in-memory storage in an {@code ArrayList} of {@code String}s.
+     */
+    public static class InMemoryContent implements Content {
+        private final ArrayList<String> paragraphs = new ArrayList<>();
```

Comment:
Since paragraphs may be inserted anywhere, I would guess that using a LinkedList should be more efficient here. Otherwise, when adding a line break somewhere in the middle of a long text, a lot of array members must be copied.

Reply:
You are right, but this is just a default implementation.  you can still implement your own `BasicTextModel.Content` and pass it to the constructor.

Do you want to see an in-memory implementation optimized for large number of paragraphs or should it better be left up to the app developers?

Reply:
Yes, we should definitely provide a very efficient implementation, because most app developers will not want to change it. We should definitely avoid another "TextArea situation", where the application breaks apart, if the text grows too big.

Reply:
I am strongly against `LinkedList` here: not only it adds 2 pointers to each element, but mostly because it provides no random access.

ArrayList should be sufficient for simple cases, and the developers are free to implement a better Content _for their use case_.  Maybe it's a skip list, or BPlusTree, or something else.

I would also like to mention that the performance issue with TextArea is not the storage (at least it is not the first order issue), but the fact that it renders all the text (in a single Text node).  RTA renders only the sliding window, which should be much faster.  Of course, RTA will have issues with very long paragraphs, but that's a different story (and is mentioned in the Non-Goals section)

https://github.com/andy-goryachev-oracle/Test/blob/main/doc/RichTextArea/RichTextArea.md#non-goals